## - Creating <font color='blue'>m18_Global_QLQ30groups65gl</font> feature based on M18_Global_QLQ30
## - Using <u>most discriminating</u> features from <font color='blue'>m0</font>,  <font color='blue'>m3</font> and <font color='blue'>m6</font>
## - SVM - Support Vector Machine
## - SMOTE and without SMOTE 

In [92]:
#importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import pyreadstat

In [93]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

In [94]:
#reading data from SPSS file
df=pd.read_spss(r"C:\Users\Chris\Desktop\Cancer_Prediction\BOUNCE_Full Dataset_m0_m3_m6_m9_m12_m15_m18_09-12-2021_ORIGINAL.sav")

In [95]:
# create a list of our conditions (breakpoint 65)
# >65  ->  0
# <65  ->  1
conditions = [(df['M18_Global_QLQ30'] >= 0) & (df['M18_Global_QLQ30'] <= 65),(df['M18_Global_QLQ30'] > 65) & (df['M18_Global_QLQ30'] <= 100)]

In [96]:
# create a list of the values we want to assign for each condition
values = [9, 8]

In [97]:
df['m18_Global_QLQ30groups65gl'] = np.select(conditions, values)

C:\Users\Chris\AppData\Local\Temp/ipykernel_2012/2152058818.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df['m18_Global_QLQ30groups65gl'] = np.select(conditions, values)


In [98]:
df

mrn  FORTH_id  M12_data      MissingM0  \
0                 1/2019 MJF       1.0       1.0  complete data   
1                10/2019 ICM       2.0       1.0  complete data   
2               100/2019 AIM       3.0       0.0  complete data   
3               101/2019 MDG       4.0       1.0  complete data   
4               103/2019 PCP       5.0       1.0  complete data   
..                       ...       ...       ...            ...   
761                  EIEO117     832.0       0.0            NaN   
762                  EIEO197     833.0       1.0            NaN   
763                  EIEO225     834.0       0.0            NaN   
764             154/2019 ACL     836.0       NaN            NaN   
765  Rabin Medical Center_50     837.0       NaN            NaN   

                   M0_MLmodels      misclasssifiedM0  misclasssifiedM0_012  \
0    used in modelling (N=609)  correctly classified  correctly classified   
1    used in modelling (N=609)  correctly classified  correctly classified   
2    used in modelling (N=609)  correctly classified  correctly classified   
3    used in modelling (N=609)  correctly classified  correctly classified   
4    used in modelling (N=609)  correctly classified  correctly classified   
..                         ...                   ...                   ...   
761                        NaN                   NaN                   NaN   
762                        NaN                   NaN                   NaN   
763                        NaN                   NaN                   NaN   
764                        NaN                   NaN                   NaN   
765                        NaN                   NaN                   NaN   

                       age.date_filter   age care_team  ...  m18_anxiety_HADS  \
0    40-70 years old within date range  66.0     CHAMP  ...          0.428571   
1    40-70 years old within date range  54.0     CHAMP  ...          0.000000   
2    40-70 years old within date range  45.0     CHAMP  ...               NaN   
3    40-70 years old within date range  70.0     CHAMP  ...          0.000000   
4    40-70 years old within date range  49.0     CHAMP  ...          1.285714   
..                                 ...   ...       ...  ...               ...   
761          outside age range or date   NaN       IEO  ...          1.000000   
762  40-70 years old within date range  43.0       IEO  ...               NaN   
763          outside age range or date   NaN       IEO  ...               NaN   
764                                NaN   NaN     CHAMP  ...          0.000000   
765                                NaN   NaN      HUJI  ...          0.142857   

    m18_depression_HADS  m18_mental_health_total_HADS  \
0              0.000000                      0.214286   
1              0.428571                      0.214286   
2                   NaN                           NaN   
3              0.000000                      0.000000   
4              1.142857                      1.214286   
..                  ...                           ...   
761            0.285714                      0.642857   
762                 NaN                           NaN   
763                 NaN                           NaN   
764            0.142857                      0.071429   
765            0.000000                      0.071429   

    m18_negative_affect_PANAS  m18_positive_affect_PANAS m18_cancer_stage  \
0                         1.2                        4.0           Stage1   
1                         1.2                        4.4          Stage 2   
2                         NaN                        NaN          Stage 3   
3                         1.0                        3.8              NaN   
4                         2.0                        3.0              NaN   
..                        ...                        ...              ...   
761                       3.2                        4.0           Stage1   
762      

In [99]:
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(0, np.NaN)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(9, 1)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(8, 0)

In [100]:
df

mrn  FORTH_id  M12_data      MissingM0  \
0                 1/2019 MJF       1.0       1.0  complete data   
1                10/2019 ICM       2.0       1.0  complete data   
2               100/2019 AIM       3.0       0.0  complete data   
3               101/2019 MDG       4.0       1.0  complete data   
4               103/2019 PCP       5.0       1.0  complete data   
..                       ...       ...       ...            ...   
761                  EIEO117     832.0       0.0            NaN   
762                  EIEO197     833.0       1.0            NaN   
763                  EIEO225     834.0       0.0            NaN   
764             154/2019 ACL     836.0       NaN            NaN   
765  Rabin Medical Center_50     837.0       NaN            NaN   

                   M0_MLmodels      misclasssifiedM0  misclasssifiedM0_012  \
0    used in modelling (N=609)  correctly classified  correctly classified   
1    used in modelling (N=609)  correctly classified  correctly classified   
2    used in modelling (N=609)  correctly classified  correctly classified   
3    used in modelling (N=609)  correctly classified  correctly classified   
4    used in modelling (N=609)  correctly classified  correctly classified   
..                         ...                   ...                   ...   
761                        NaN                   NaN                   NaN   
762                        NaN                   NaN                   NaN   
763                        NaN                   NaN                   NaN   
764                        NaN                   NaN                   NaN   
765                        NaN                   NaN                   NaN   

                       age.date_filter   age care_team  ...  m18_anxiety_HADS  \
0    40-70 years old within date range  66.0     CHAMP  ...          0.428571   
1    40-70 years old within date range  54.0     CHAMP  ...          0.000000   
2    40-70 years old within date range  45.0     CHAMP  ...               NaN   
3    40-70 years old within date range  70.0     CHAMP  ...          0.000000   
4    40-70 years old within date range  49.0     CHAMP  ...          1.285714   
..                                 ...   ...       ...  ...               ...   
761          outside age range or date   NaN       IEO  ...          1.000000   
762  40-70 years old within date range  43.0       IEO  ...               NaN   
763          outside age range or date   NaN       IEO  ...               NaN   
764                                NaN   NaN     CHAMP  ...          0.000000   
765                                NaN   NaN      HUJI  ...          0.142857   

    m18_depression_HADS  m18_mental_health_total_HADS  \
0              0.000000                      0.214286   
1              0.428571                      0.214286   
2                   NaN                           NaN   
3              0.000000                      0.000000   
4              1.142857                      1.214286   
..                  ...                           ...   
761            0.285714                      0.642857   
762                 NaN                           NaN   
763                 NaN                           NaN   
764            0.142857                      0.071429   
765            0.000000                      0.071429   

    m18_negative_affect_PANAS  m18_positive_affect_PANAS m18_cancer_stage  \
0                         1.2                        4.0           Stage1   
1                         1.2                        4.4          Stage 2   
2                         NaN                        NaN          Stage 3   
3                         1.0                        3.8              NaN   
4                         2.0                        3.0              NaN   
..                        ...                        ...              ...   
761                       3.2                        4.0           Stage1   
762      

<b>0</b> -> <font color='blue'>good QoL</font> 
<br>
<b>1</b> -> <font color='red'>poor QoL</font> 

## -------------- (end of <i>Creating m18_Global_QLQ30groups65gl</i>) --------------

-------

## Dataset which includes only <font color='blue'>m0</font>, <font color='blue'>m3</font> and <font color='blue'>m6</font> feautres 

In [101]:
df = df[['M0_employment_status','M0_drinking','M0_diet','M0_MedDiet','M0_SpecialDiet','M0_weight','M0_bmi',
         'M0_obese','M0_exercise','M0_life_events_1.2','M0_sick_leave_days','M0_extraversion_TIPI',
         'M0_aggreableness_TIPI','M0_conscientiousness_TIPI','M0_emot_stab_TIPI','M0_openess_TIPI',
         'M0_optimism_LOT','M0_comprehensibility_SOC','M0_manageability_SOC','M0_meaningfulness_SOC',
         'M0_Forward_PACT','M0_Trauma_PACT','M0_Flexibility_PACT','M0_self_blame_CERQ','M0_other_blame_CERQ',
         'M0_rumination_CERQ','M0_catastrophizing_CERQ','M0_perspective_CERQ','M0_pos_refus_CERQ',
         'M0_pos_reapp_CERQ','M0_acceptance_CERQ','M0_planning_CERQ','M0_negative_overall_CERQ',
         'M0_positive_overall_CERQ','M0_mindfulness_MAAS','M0_resilience_CDRISC','M0_coping_with_cancer_CBI',
         'M0_fear_of_recur_FCRI','M0_anxiety_HADS','M0_depression_HADS','M0_mental_health_total_HADS',
         'M0_positive_affect_PANAS','M0_negative_affect_PANAS','M0_Global_QLQ30','M0_Phys_Fun_QLQ30',
         'M0_Role_Fun_QLQ30','M0_Emot_Fun_QLQ30','M0_Cogn_Fun_QLQ30','M0_Soc_Fun_QLQ30','M0_Fatigue_QLQ30',
         'M0_Nausea_QLQ30','M0_Pain_QLQ30','M0_Dyspnoea_QLQ30','M0_Insomnia_QLQ30','M0_Apetite_QLQ30',
         'M0_Constipation_QLQ30','M0_Diarrhoea_QLQ30','M0_Financial_QLQ30','M0_Body_Image_BR23',
         'M0_Side_Effects_BR23','M0_Breast_Symptoms_BR23','M0_Arm_Symptoms_BR23','M0_Upset_Hair_Image_BR23',
         'M0_Future_Persp_Image_BR23','M0_Sex_Funct_BR23','M0_Sex_Enjoy_BR23','M0_performancestatus',
         'm0_menopausalstatuspre','M0_surgery','M0_chemo','M0_radio','M0_psychotropics',
         'm0_Global_QLQ30groups65gl','m0_Phys_Fun_QLQ30groups80gl','diagnosis_M0_delay','m3_life_events_1.2',
         'm3_sick_leave_days','m3_employer_support','m3_domestic_help',
         'm3_perceived_suppport_1_item','m3_nccn_distress_thermometer',
         'm3_general_se_1_item','m3_rate_how_much_your_faith_including_religiosity_and_spiri',
         'M3_instumental_support','M3_emotional_support','M3_social_support_total','M3_PTGI_relating_to_others',
         'M3_PTGI_new_possibilities','M3_PTGI_personal_strength','M3_PTGI_spiritual_change','M3_PTGI_appreciation_of_life',
         'M3_PTGI_total_score','M3_negative_affect_PANAS','M3_positive_affect_PANAS','M3_anxiety_HADS',
         'M3_depression_HADS','M3_mental_health_total_HADS','M3_Global_QLQ30','M3_Phys_Fun_QLQ30',
         'M3_Role_Fun_QLQ30','M3_Emot_Fun_QLQ30','M3_Cogn_Fun_QLQ30','M3_Soc_Fun_QLQ30','M3_Fatigue_QLQ30',
         'M3_Nausea_QLQ30','M3_Pain_QLQ30','M3_Dyspnoea_QLQ30','M3_Insomnia_QLQ30','M3_Apetite_QLQ30',
         'M3_Constipation_QLQ30','M3_Diarrhoea_QLQ30','M3_Financial_QLQ30','M3_Body_Image_m3_br23',
         'M3_Side_Effects_m3_br23','M3_Breast_Symptoms_m3_br23','M3_Arm_Symptoms_m3_br23','M3_Upset_Hair_Image_m3_br23',
         'M3_Future_Persp_Image_m3_br23','M3_Sex_Funct_m3_br23','M3_Sex_Enjoy_m3_br23','m3_performancestatus',
         'M3_surgery','M3_chemo','M3_radio','M3_psychotropics','M3_Mental_health_support',
         'm3_illness_events0.1.2','m3_Global_QLQ30groups65gl','m3_Phys_Fun_QLQ30groups80gl','M6_employment_status',
         'm6_weight','M6_bmi','M6_exercise','M6_life_events_1.2','M6_sick_leave_days','M6_well_being_activities',
         'M6_domestic_help','M6_partner_reducework','M6_perceived_suppport_1_item','M6_distress_thermometer',
         'M6_general_se_1_item','M6_how_much_are_you_back_to_yourself','M6_coping_with_cancer_CBI',
         'M6_fear_of_recur_FCRI','M6_ptsd_PCL5','M6_IPQ_timeline','M6_negative_affect_PANAS','M6_positive_affect_PANAS',
         'M6_anxiety_HADS','M6_depression_HADS','M6_mental_health_total_HADS','M6_Global_QLQ30','M6_Phys_Fun_QLQ30',
         'M6_Role_Fun_QLQ30','M6_Emot_Fun_QLQ30','M6_Cogn_Fun_QLQ30','M6_Soc_Fun_QLQ30','M6_Fatigue_QLQ30',
         'M6_Nausea_QLQ30','M6_Pain_QLQ30','M6_Dyspnoea_QLQ30','M6_Insomnia_QLQ30','M6_Apetite_QLQ30','M6_Constipation_QLQ30',
         'M6_Diarrhoea_QLQ30','M6_Financial_QLQ30','M6_Body_Image_br23','M6_Side_Effects_br23','M6_Breast_Symptoms_br23',
         'M6_Arm_Symptoms_br23','M6_Upset_Hair_Image_br23','M6_Future_Persp_Image_br23','M6_Sex_Funct_br23',
         'M6_Sex_Enjoy_br23','m6_performancestatus','M6_surgery','M6_chemo','M6_radio',
         'M6_psychotropics','M6_number_consultations','M6_Mental_health_support','m6_illness_events0.1.2',
         'M6_drugs0_1','m6_Global_QLQ30groups65gl','m6_Phys_Fun_QLQ30groups80gl','m18_Global_QLQ30groups65gl']]

In [102]:
#check the category features
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 186 columns):
 #    Column                                                       Dtype   
---   ------                                                       -----   
 0    M0_employment_status                                         category
 1    M0_drinking                                                  category
 2    M0_diet                                                      category
 3    M0_MedDiet                                                   category
 4    M0_SpecialDiet                                               category
 5    M0_weight                                                    float64 
 6    M0_bmi                                                       float64 
 7    M0_obese                                                     category
 8    M0_exercise                                                  category
 9    M0_life_events_1.2                                  

In [103]:
#convert category to numeric value
from sklearn.preprocessing import LabelEncoder
labelencoder_Y = LabelEncoder()
df.iloc[:,0] = labelencoder_Y.fit_transform(df.iloc[:,0].values)
df.iloc[:,1] = labelencoder_Y.fit_transform(df.iloc[:,1].values)
df.iloc[:,2] = labelencoder_Y.fit_transform(df.iloc[:,2].values)
df.iloc[:,3] = labelencoder_Y.fit_transform(df.iloc[:,3].values)
df.iloc[:,4] = labelencoder_Y.fit_transform(df.iloc[:,4].values)
df.iloc[:,7] = labelencoder_Y.fit_transform(df.iloc[:,7].values)
df.iloc[:,8] = labelencoder_Y.fit_transform(df.iloc[:,8].values)
df.iloc[:,9] = labelencoder_Y.fit_transform(df.iloc[:,9].values)
df.iloc[:,67] = labelencoder_Y.fit_transform(df.iloc[:,67].values)
df.iloc[:,71] = labelencoder_Y.fit_transform(df.iloc[:,71].values)
df.iloc[:,72] = labelencoder_Y.fit_transform(df.iloc[:,72].values)
df.iloc[:,73] = labelencoder_Y.fit_transform(df.iloc[:,73].values)
df.iloc[:,75] = labelencoder_Y.fit_transform(df.iloc[:,75].values)
df.iloc[:,77] = labelencoder_Y.fit_transform(df.iloc[:,77].values)
df.iloc[:,78] = labelencoder_Y.fit_transform(df.iloc[:,78].values)
df.iloc[:,124] = labelencoder_Y.fit_transform(df.iloc[:,124].values)
df.iloc[:,125] = labelencoder_Y.fit_transform(df.iloc[:,125].values)
df.iloc[:,126] = labelencoder_Y.fit_transform(df.iloc[:,126].values)
df.iloc[:,127] = labelencoder_Y.fit_transform(df.iloc[:,127].values)
df.iloc[:,128] = labelencoder_Y.fit_transform(df.iloc[:,128].values)
df.iloc[:,129] = labelencoder_Y.fit_transform(df.iloc[:,129].values)
df.iloc[:,132] = labelencoder_Y.fit_transform(df.iloc[:,132].values)
df.iloc[:,133] = labelencoder_Y.fit_transform(df.iloc[:,133].values)
df.iloc[:,135] = labelencoder_Y.fit_transform(df.iloc[:,135].values)
df.iloc[:,136] = labelencoder_Y.fit_transform(df.iloc[:,136].values)
df.iloc[:,137] = labelencoder_Y.fit_transform(df.iloc[:,137].values)
df.iloc[:,178] = labelencoder_Y.fit_transform(df.iloc[:,178].values)
df.iloc[:,180] = labelencoder_Y.fit_transform(df.iloc[:,180].values)
df.iloc[:,181] = labelencoder_Y.fit_transform(df.iloc[:,181].values)
df.iloc[:,182] = labelencoder_Y.fit_transform(df.iloc[:,182].values)
df.iloc[:,183] = labelencoder_Y.fit_transform(df.iloc[:,183].values)
df.iloc[:,184] = labelencoder_Y.fit_transform(df.iloc[:,184].values)

c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [104]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                       0            1        1           1               0   
1                       0            1        1           1               0   
2                       4            1        1           1               0   
3                       2            2        1           1               0   
4                       2            1        0           0               0   
..                    ...          ...      ...         ...             ...   
761                     6            3        3           2               2   
762                     4            1        0           0               0   
763                     6            3        3           2               2   
764                     6            3        3           2               2   
765                     6            3        3           2               2   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1.2  ...  \
0         77.0  28.628792         0            1                   1  ...   
1         52.0  22.506925         0            2                   1  ...   
2         68.0  27.942143         0            2                   1  ...   
3         53.0  24.196494         0            2                   1  ...   
4         52.0  19.100092         0            1                   1  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN         2            3                   1  ...   
762       54.0  19.132653         0            1                   0  ...   
763        NaN        NaN         2            3                   1  ...   
764        NaN        NaN         2            3                   2  ...   
765        NaN        NaN         2            3                   2  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0         0.0       0.0                 0                      NaN   
1         0.0       0.0                 0                      NaN   
2         NaN       NaN                 0                      NaN   
3         0.0       0.0                 0                      NaN   
4         0.0       0.0                 0                      1.0   
..        ...       ...               ...                      ...   
761       NaN       NaN                 0                      NaN   
762       0.0       0.0                 0                      NaN   
763       0.0       0.0                 0                      NaN   
764       NaN       NaN                 2                      NaN   
765       NaN       NaN                 2                      NaN   

     M6_Mental_health_support  m6_illness_events0.1.2  M6_drugs0_1  \
0                           0                       1            0   
1                           0                       2            0   
2                           2                       3            2   
3                           2                       1            0   
4                           0                       2            0   
..                        ...                     ...          ...   
761                         2                       3            0   
762                         0                       3            0   
763                         0                       3            0   
764                         2                       3            2   
765                         2                       3            2   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                            0                            1   
1                            1                            0   
2                            2                            2   
3                            0                            0   
4                            0                            1   
..                         ...                          ...   
761     

In [105]:
#rename 3 columns which include dots (.) in their names 
df = df.rename(columns={'M0_life_events_1.2': 'M0_life_events_1_2', 'm3_life_events_1.2': 'm3_life_events_1_2', 'm3_illness_events0.1.2':'m3_illness_events0_1_2'})
df = df.rename(columns={'M6_life_events_1.2': 'M6_life_events_1_2', 'm6_illness_events0.1.2':'m6_illness_events0_1_2'})

In [106]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 186 columns):
 #    Column                                                       Dtype  
---   ------                                                       -----  
 0    M0_employment_status                                         int32  
 1    M0_drinking                                                  int32  
 2    M0_diet                                                      int32  
 3    M0_MedDiet                                                   int32  
 4    M0_SpecialDiet                                               int32  
 5    M0_weight                                                    float64
 6    M0_bmi                                                       float64
 7    M0_obese                                                     int32  
 8    M0_exercise                                                  int32  
 9    M0_life_events_1_2                                           in

In [107]:
#convert corresponding int32 feature type to float64
df['M0_employment_status'] = df.M0_employment_status.astype(float)
df['M0_drinking'] = df.M0_drinking.astype(float)
df['M0_diet'] = df.M0_diet.astype(float)
df['M0_MedDiet'] = df.M0_MedDiet.astype(float)
df['M0_SpecialDiet'] = df.M0_SpecialDiet.astype(float)
df['M0_obese'] = df.M0_obese.astype(float)
df['M0_exercise'] = df.M0_exercise.astype(float)
df['M0_life_events_1_2'] = df.M0_life_events_1_2.astype(float)
df['m0_menopausalstatuspre'] = df.m0_menopausalstatuspre.astype(float)
df['M0_psychotropics'] = df.M0_psychotropics.astype(float)
df['m0_Global_QLQ30groups65gl'] = df.m0_Global_QLQ30groups65gl.astype(float)
df['m0_Phys_Fun_QLQ30groups80gl'] = df.m0_Phys_Fun_QLQ30groups80gl.astype(float)
df['m3_life_events_1_2'] = df.m3_life_events_1_2.astype(float)
df['m3_employer_support'] = df.m3_employer_support.astype(float)
df['m3_domestic_help'] = df.m3_domestic_help.astype(float)
df['M3_psychotropics'] = df.M3_psychotropics.astype(float)
df['M3_Mental_health_support'] = df.M3_Mental_health_support.astype(float)
df['m3_illness_events0_1_2'] = df.m3_illness_events0_1_2.astype(float)
df['m3_Global_QLQ30groups65gl'] = df.m3_Global_QLQ30groups65gl.astype(float)
df['m3_Phys_Fun_QLQ30groups80gl'] = df.m3_Phys_Fun_QLQ30groups80gl.astype(float)
df['M6_employment_status'] = df.M6_employment_status.astype(float)
df['M6_exercise'] = df.M6_exercise.astype(float)
df['M6_life_events_1_2'] = df.M6_life_events_1_2.astype(float)
df['M6_well_being_activities'] = df.M6_well_being_activities.astype(float)
df['M6_domestic_help'] = df.M6_domestic_help.astype(float)
df['M6_partner_reducework'] = df.M6_partner_reducework.astype(float)
df['M6_psychotropics'] = df.M6_psychotropics.astype(float)
df['M6_Mental_health_support'] = df.M6_Mental_health_support.astype(float)
df['m6_illness_events0_1_2'] = df.m6_illness_events0_1_2.astype(float)
df['M6_drugs0_1'] = df.M6_drugs0_1.astype(float)
df['m6_Global_QLQ30groups65gl'] = df.m6_Global_QLQ30groups65gl.astype(float)
df['m6_Phys_Fun_QLQ30groups80gl'] = df.m6_Phys_Fun_QLQ30groups80gl.astype(float)

In [108]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 186 columns):
 #    Column                                                       Dtype  
---   ------                                                       -----  
 0    M0_employment_status                                         float64
 1    M0_drinking                                                  float64
 2    M0_diet                                                      float64
 3    M0_MedDiet                                                   float64
 4    M0_SpecialDiet                                               float64
 5    M0_weight                                                    float64
 6    M0_bmi                                                       float64
 7    M0_obese                                                     float64
 8    M0_exercise                                                  float64
 9    M0_life_events_1_2                                           fl

In [109]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   6.0          3.0      3.0         2.0             2.0   
762                   4.0          1.0      0.0         0.0             0.0   
763                   6.0          3.0      3.0         2.0             2.0   
764                   6.0          3.0      3.0         2.0             2.0   
765                   6.0          3.0      3.0         2.0             2.0   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0         77.0  28.628792       0.0          1.0                 1.0  ...   
1         52.0  22.506925       0.0          2.0                 1.0  ...   
2         68.0  27.942143       0.0          2.0                 1.0  ...   
3         53.0  24.196494       0.0          2.0                 1.0  ...   
4         52.0  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN       2.0          3.0                 1.0  ...   
762       54.0  19.132653       0.0          1.0                 0.0  ...   
763        NaN        NaN       2.0          3.0                 1.0  ...   
764        NaN        NaN       2.0          3.0                 2.0  ...   
765        NaN        NaN       2.0          3.0                 2.0  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0         0.0       0.0               0.0                      NaN   
1         0.0       0.0               0.0                      NaN   
2         NaN       NaN               0.0                      NaN   
3         0.0       0.0               0.0                      NaN   
4         0.0       0.0               0.0                      1.0   
..        ...       ...               ...                      ...   
761       NaN       NaN               0.0                      NaN   
762       0.0       0.0               0.0                      NaN   
763       0.0       0.0               0.0                      NaN   
764       NaN       NaN               2.0                      NaN   
765       NaN       NaN               2.0                      NaN   

     M6_Mental_health_support  m6_illness_events0_1_2  M6_drugs0_1  \
0                         0.0                     1.0          0.0   
1                         0.0                     2.0          0.0   
2                         2.0                     3.0          2.0   
3                         2.0                     1.0          0.0   
4                         0.0                     2.0          0.0   
..                        ...                     ...          ...   
761                       2.0                     3.0          0.0   
762                       0.0                     3.0          0.0   
763                       0.0                     3.0          0.0   
764                       2.0                     3.0          2.0   
765                       2.0                     3.0          2.0   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          1.0                          0.0   
2                          2.0                          2.0   
3                          0.0                          0.0   
4                          0.0                          1.0   
..                         ...                          ...   
761     

In [110]:
#convert corresponding numeric value to NaN
df['M0_employment_status']=df['M0_employment_status'].replace(6, np.NaN)
df['M0_drinking']=df['M0_drinking'].replace(3, np.NaN)
df['M0_diet']=df['M0_diet'].replace(3, np.NaN)
df['M0_MedDiet']=df['M0_MedDiet'].replace(2, np.NaN)
df['M0_SpecialDiet']=df['M0_SpecialDiet'].replace(2, np.NaN)
df['M0_obese']=df['M0_obese'].replace(2, np.NaN)
df['M0_exercise']=df['M0_exercise'].replace(3, np.NaN)
df['M0_life_events_1_2']=df['M0_life_events_1_2'].replace(2, np.NaN)
df['m0_menopausalstatuspre']=df['m0_menopausalstatuspre'].replace(3, np.NaN)
df['M0_psychotropics']=df['M0_psychotropics'].replace(2, np.NaN)
df['m0_Global_QLQ30groups65gl']=df['m0_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m0_Phys_Fun_QLQ30groups80gl']=df['m0_Phys_Fun_QLQ30groups80gl'].replace(2, np.NaN)
df['m3_life_events_1_2']=df['m3_life_events_1_2'].replace(2, np.NaN)
df['m3_employer_support']=df['m3_employer_support'].replace(2, np.NaN)
df['m3_domestic_help']=df['m3_domestic_help'].replace(2, np.NaN)
df['M3_psychotropics']=df['M3_psychotropics'].replace(3, np.NaN)
df['M3_Mental_health_support']=df['M3_Mental_health_support'].replace(2, np.NaN)
df['m3_illness_events0_1_2']=df['m3_illness_events0_1_2'].replace(3, np.NaN)
df['m3_Global_QLQ30groups65gl']=df['m3_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m3_Phys_Fun_QLQ30groups80gl']=df['m3_Phys_Fun_QLQ30groups80gl'].replace(2, np.NaN)
df['M6_employment_status'] = df['M6_employment_status'].replace(6, np.NaN)
df['M6_exercise'] = df['M6_exercise'].replace(3, np.NaN)
df['M6_life_events_1_2'] = df['M6_life_events_1_2'].replace(2, np.NaN)
df['M6_well_being_activities'] = df['M6_well_being_activities'].replace(2, np.NaN)
df['M6_domestic_help'] = df['M6_domestic_help'].replace(2, np.NaN)
df['M6_partner_reducework'] = df['M6_partner_reducework'].replace(2, np.NaN)
df['M6_psychotropics'] = df['M6_psychotropics'].replace(2, np.NaN)
df['M6_Mental_health_support'] = df['M6_Mental_health_support'].replace(2, np.NaN)
df['m6_illness_events0_1_2'] = df['m6_illness_events0_1_2'].replace(3, np.NaN)
df['M6_drugs0_1'] = df['M6_drugs0_1'].replace(2, np.NaN)
df['m6_Global_QLQ30groups65gl'] = df['m6_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m6_Phys_Fun_QLQ30groups80gl'] = df['m6_Phys_Fun_QLQ30groups80gl'].replace(2, np.NaN)

In [111]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   NaN          NaN      NaN         NaN             NaN   
762                   4.0          1.0      0.0         0.0             0.0   
763                   NaN          NaN      NaN         NaN             NaN   
764                   NaN          NaN      NaN         NaN             NaN   
765                   NaN          NaN      NaN         NaN             NaN   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0         77.0  28.628792       0.0          1.0                 1.0  ...   
1         52.0  22.506925       0.0          2.0                 1.0  ...   
2         68.0  27.942143       0.0          2.0                 1.0  ...   
3         53.0  24.196494       0.0          2.0                 1.0  ...   
4         52.0  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN       NaN          NaN                 1.0  ...   
762       54.0  19.132653       0.0          1.0                 0.0  ...   
763        NaN        NaN       NaN          NaN                 1.0  ...   
764        NaN        NaN       NaN          NaN                 NaN  ...   
765        NaN        NaN       NaN          NaN                 NaN  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0         0.0       0.0               0.0                      NaN   
1         0.0       0.0               0.0                      NaN   
2         NaN       NaN               0.0                      NaN   
3         0.0       0.0               0.0                      NaN   
4         0.0       0.0               0.0                      1.0   
..        ...       ...               ...                      ...   
761       NaN       NaN               0.0                      NaN   
762       0.0       0.0               0.0                      NaN   
763       0.0       0.0               0.0                      NaN   
764       NaN       NaN               NaN                      NaN   
765       NaN       NaN               NaN                      NaN   

     M6_Mental_health_support  m6_illness_events0_1_2  M6_drugs0_1  \
0                         0.0                     1.0          0.0   
1                         0.0                     2.0          0.0   
2                         NaN                     NaN          NaN   
3                         NaN                     1.0          0.0   
4                         0.0                     2.0          0.0   
..                        ...                     ...          ...   
761                       NaN                     NaN          0.0   
762                       0.0                     NaN          0.0   
763                       0.0                     NaN          0.0   
764                       NaN                     NaN          NaN   
765                       NaN                     NaN          NaN   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          1.0                          0.0   
2                          NaN                          NaN   
3                          0.0                          0.0   
4                          0.0                          1.0   
..                         ...                          ...   
761     

------------------

<font color='blue'><b>MICE Data Imoutation</b></font>

In [21]:
imputer = IterativeImputer(imputation_order='ascending',max_iter=10,random_state=42,n_nearest_features=None)

In [22]:
df = imputer.fit_transform(df)

c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\sklearn\impute\_iterative.py:699: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(


In [23]:
df

array([[0.        , 1.        , 1.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 1.        , 1.        , ..., 1.        , 0.        ,
        0.        ],
       [4.        , 1.        , 1.        , ..., 0.10488535, 0.37161837,
        0.54429797],
       ...,
       [1.53197826, 1.60108087, 0.7602016 , ..., 0.        , 1.        ,
        0.1454644 ],
       [1.50558957, 1.19502978, 0.76244958, ..., 0.22714458, 0.40317828,
        0.        ],
       [1.60875148, 1.17069928, 0.72448332, ..., 0.25384964, 0.408191  ,
        1.        ]])

In [24]:
#Convert Array to Dataframe and name columns
df=pd.DataFrame(df,columns = ['M0_employment_status','M0_drinking','M0_diet','M0_MedDiet','M0_SpecialDiet','M0_weight','M0_bmi',
         'M0_obese','M0_exercise','M0_life_events_1_2','M0_sick_leave_days','M0_extraversion_TIPI',
         'M0_aggreableness_TIPI','M0_conscientiousness_TIPI','M0_emot_stab_TIPI','M0_openess_TIPI',
         'M0_optimism_LOT','M0_comprehensibility_SOC','M0_manageability_SOC','M0_meaningfulness_SOC',
         'M0_Forward_PACT','M0_Trauma_PACT','M0_Flexibility_PACT','M0_self_blame_CERQ','M0_other_blame_CERQ',
         'M0_rumination_CERQ','M0_catastrophizing_CERQ','M0_perspective_CERQ','M0_pos_refus_CERQ',
         'M0_pos_reapp_CERQ','M0_acceptance_CERQ','M0_planning_CERQ','M0_negative_overall_CERQ',
         'M0_positive_overall_CERQ','M0_mindfulness_MAAS','M0_resilience_CDRISC','M0_coping_with_cancer_CBI',
         'M0_fear_of_recur_FCRI','M0_anxiety_HADS','M0_depression_HADS','M0_mental_health_total_HADS',
         'M0_positive_affect_PANAS','M0_negative_affect_PANAS','M0_Global_QLQ30','M0_Phys_Fun_QLQ30',
         'M0_Role_Fun_QLQ30','M0_Emot_Fun_QLQ30','M0_Cogn_Fun_QLQ30','M0_Soc_Fun_QLQ30','M0_Fatigue_QLQ30',
         'M0_Nausea_QLQ30','M0_Pain_QLQ30','M0_Dyspnoea_QLQ30','M0_Insomnia_QLQ30','M0_Apetite_QLQ30',
         'M0_Constipation_QLQ30','M0_Diarrhoea_QLQ30','M0_Financial_QLQ30','M0_Body_Image_BR23',
         'M0_Side_Effects_BR23','M0_Breast_Symptoms_BR23','M0_Arm_Symptoms_BR23','M0_Upset_Hair_Image_BR23',
         'M0_Future_Persp_Image_BR23','M0_Sex_Funct_BR23','M0_Sex_Enjoy_BR23','M0_performancestatus',
         'm0_menopausalstatuspre','M0_surgery','M0_chemo','M0_radio','M0_psychotropics',
         'm0_Global_QLQ30groups65gl','m0_Phys_Fun_QLQ30groups80gl','diagnosis_M0_delay','m3_life_events_1_2',
         'm3_sick_leave_days','m3_employer_support','m3_domestic_help',
         'm3_perceived_suppport_1_item','m3_nccn_distress_thermometer',
         'm3_general_se_1_item','m3_rate_how_much_your_faith_including_religiosity_and_spiri',
         'M3_instumental_support','M3_emotional_support','M3_social_support_total','M3_PTGI_relating_to_others',
         'M3_PTGI_new_possibilities','M3_PTGI_personal_strength','M3_PTGI_spiritual_change','M3_PTGI_appreciation_of_life',
         'M3_PTGI_total_score','M3_negative_affect_PANAS','M3_positive_affect_PANAS','M3_anxiety_HADS',
         'M3_depression_HADS','M3_mental_health_total_HADS','M3_Global_QLQ30','M3_Phys_Fun_QLQ30',
         'M3_Role_Fun_QLQ30','M3_Emot_Fun_QLQ30','M3_Cogn_Fun_QLQ30','M3_Soc_Fun_QLQ30','M3_Fatigue_QLQ30',
         'M3_Nausea_QLQ30','M3_Pain_QLQ30','M3_Dyspnoea_QLQ30','M3_Insomnia_QLQ30','M3_Apetite_QLQ30',
         'M3_Constipation_QLQ30','M3_Diarrhoea_QLQ30','M3_Financial_QLQ30','M3_Body_Image_m3_br23',
         'M3_Side_Effects_m3_br23','M3_Breast_Symptoms_m3_br23','M3_Arm_Symptoms_m3_br23','M3_Upset_Hair_Image_m3_br23',
         'M3_Future_Persp_Image_m3_br23','M3_Sex_Funct_m3_br23','M3_Sex_Enjoy_m3_br23','m3_performancestatus',
         'M3_surgery','M3_chemo','M3_radio','M3_psychotropics','M3_Mental_health_support',
         'm3_illness_events0_1_2','m3_Global_QLQ30groups65gl','m3_Phys_Fun_QLQ30groups80gl','M6_employment_status',
         'm6_weight','M6_bmi','M6_exercise','M6_life_events_1_2','M6_sick_leave_days','M6_well_being_activities',
         'M6_domestic_help','M6_partner_reducework','M6_perceived_suppport_1_item','M6_distress_thermometer',
         'M6_general_se_1_item','M6_how_much_are_you_back_to_yourself','M6_coping_with_cancer_CBI',
         'M6_fear_of_recur_FCRI','M6_ptsd_PCL5','M6_IPQ_timeline','M6_negative_affect_PANAS','M6_positive_affect_PANAS',
         'M6_anxiety_HADS','M6_depression_HADS','M6_mental_health_total_HADS','M6_Global_QLQ30','M6_Phys_Fun_QLQ30',
         'M6_Role_Fun_QLQ30','M6_Emot_Fun_QLQ30','M6_Cogn_Fun_QLQ30','M6_Soc_Fun_QLQ30','M6_Fatigue_QLQ30',
         'M6_Nausea_QLQ30','M6_Pain_QLQ30','M6_Dyspnoea_QLQ30','M6_Insomnia_QLQ30','M6_Apetite_QLQ30','M6_Constipation_QLQ30',
         'M6_Diarrhoea_QLQ30','M6_Financial_QLQ30','M6_Body_Image_br23','M6_Side_Effects_br23','M6_Breast_Symptoms_br23',
         'M6_Arm_Symptoms_br23','M6_Upset_Hair_Image_br23','M6_Future_Persp_Image_br23','M6_Sex_Funct_br23',
         'M6_Sex_Enjoy_br23','m6_performancestatus','M6_surgery','M6_chemo','M6_radio',
         'M6_psychotropics','M6_number_consultations','M6_Mental_health_support','m6_illness_events0_1_2',
         'M6_drugs0_1','m6_Global_QLQ30groups65gl','m6_Phys_Fun_QLQ30groups80gl','m18_Global_QLQ30groups65gl'])

In [25]:
df

M0_employment_status  M0_drinking   M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                0.000000     1.000000  1.000000    1.000000        0.000000   
1                0.000000     1.000000  1.000000    1.000000        0.000000   
2                4.000000     1.000000  1.000000    1.000000        0.000000   
3                2.000000     2.000000  1.000000    1.000000        0.000000   
4                2.000000     1.000000  0.000000    0.000000        0.000000   
..                    ...          ...       ...         ...             ...   
761              1.615491     1.362700  0.797484    0.529285        0.285870   
762              4.000000     1.000000  0.000000    0.000000        0.000000   
763              1.531978     1.601081  0.760202    0.602902        0.178817   
764              1.505590     1.195030  0.762450    0.634501        0.130337   
765              1.608751     1.170699  0.724483    0.656151        0.066613   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792  0.000000     1.000000            1.000000  ...   
1    52.000000  22.506925  0.000000     2.000000            1.000000  ...   
2    68.000000  27.942143  0.000000     2.000000            1.000000  ...   
3    53.000000  24.196494  0.000000     2.000000            1.000000  ...   
4    52.000000  19.100092  0.000000     1.000000            1.000000  ...   
..         ...        ...       ...          ...                 ...  ...   
761  69.193991  25.778712  0.370289     1.185536            1.000000  ...   
762  54.000000  19.132653  0.000000     1.000000            0.000000  ...   
763  67.962041  25.439019  0.143085     1.006077            1.000000  ...   
764  68.499715  25.584960  0.148111     1.028714            0.634544  ...   
765  68.416958  25.564308  0.175523     1.028082            0.635781  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0    0.000000  0.000000          0.000000                 0.236035   
1    0.000000  0.000000          0.000000                 4.742909   
2    0.048191  0.087855          0.000000                 5.852626   
3    0.000000  0.000000          0.000000                 2.672416   
4    0.000000  0.000000          0.000000                 1.000000   
..        ...       ...               ...                      ...   
761  0.074925 -0.003216          0.000000                 4.272397   
762  0.000000  0.000000          0.000000                 6.702640   
763  0.000000  0.000000          0.000000                 5.093694   
764  0.065328  0.111087          0.180132                 4.345900   
765  0.064508  0.119566          0.182419                 4.373343   

     M6_Mental_health_support  m6_illness_events0_1_2  M6_drugs0_1  \
0                    0.000000                1.000000     0.000000   
1                    0.000000                2.000000     0.000000   
2                    0.032466                1.435106     0.100807   
3                    0.026747                1.000000     0.000000   
4                    0.000000                2.000000     0.000000   
..                        ...                     ...          ...   
761                  0.163941                1.443145     0.000000   
762                  0.000000                1.562925     0.000000   
763                  0.000000                1.649863     0.000000   
764                  0.151169                1.478833     0.062721   
765                  0.153522                1.483606     0.064308   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                     0.000000                     1.000000   
1                     1.000000                     0.000000   
2                     0.104885                     0.371618   
3                     0.000000                     0.000000   
4                     0.000000                     1.000000   
..                         ...                          ...

<b>- roundup values</b>  -

In [26]:
df.M0_employment_status = df.M0_employment_status.round()
df.M0_drinking = df.M0_drinking.round()
df.M0_diet = df.M0_diet.round()
df.M0_MedDiet = df.M0_MedDiet.round()
df.M0_SpecialDiet = df.M0_SpecialDiet.round()
df.M0_obese = df.M0_obese.round()
df.M0_exercise = df.M0_exercise.round()
df.M0_life_events_1_2 = df.M0_life_events_1_2.round()
df.m0_menopausalstatuspre = df.m0_menopausalstatuspre.round()
df.M0_psychotropics = df.M0_psychotropics.round()
df.m0_Global_QLQ30groups65gl = df.m0_Global_QLQ30groups65gl.round()
df.m0_Phys_Fun_QLQ30groups80gl = df.m0_Phys_Fun_QLQ30groups80gl.round()
df.m3_life_events_1_2 = df.m3_life_events_1_2.round()
df.m3_employer_support = df.m3_employer_support.round()
df.m3_domestic_help = df.m3_domestic_help.round()
df.M3_psychotropics = df.M3_psychotropics.round()
df.M3_Mental_health_support = df.M3_Mental_health_support.round()
df.m3_illness_events0_1_2 = df.m3_illness_events0_1_2.round()
df.m3_Global_QLQ30groups65gl = df.m3_Global_QLQ30groups65gl.round()
df.m3_Phys_Fun_QLQ30groups80gl = df.m3_Phys_Fun_QLQ30groups80gl.round()
df.M6_employment_status = df.M6_employment_status.round()
df.M6_exercise = df.M6_exercise.round()
df.M6_life_events_1_2 = df.M6_life_events_1_2.round()
df.M6_well_being_activities = df.M6_well_being_activities.round()
df.M6_domestic_help = df.M6_domestic_help.round()
df.M6_partner_reducework = df.M6_partner_reducework.round()
df.M6_psychotropics = df.M6_psychotropics.round()
df.M6_Mental_health_support = df.M6_Mental_health_support.round()
df.m6_illness_events0_1_2 = df.m6_illness_events0_1_2.round()
df.M6_drugs0_1 = df.M6_drugs0_1.round()
df.m6_Global_QLQ30groups65gl = df.m6_Global_QLQ30groups65gl.round()
df.m6_Phys_Fun_QLQ30groups80gl = df.m6_Phys_Fun_QLQ30groups80gl.round()
df.m18_Global_QLQ30groups65gl = df.m18_Global_QLQ30groups65gl.round()

In [27]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   2.0          1.0      1.0         1.0             0.0   
762                   4.0          1.0      0.0         0.0             0.0   
763                   2.0          2.0      1.0         1.0             0.0   
764                   2.0          1.0      1.0         1.0             0.0   
765                   2.0          1.0      1.0         1.0             0.0   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792       0.0          1.0                 1.0  ...   
1    52.000000  22.506925       0.0          2.0                 1.0  ...   
2    68.000000  27.942143       0.0          2.0                 1.0  ...   
3    53.000000  24.196494       0.0          2.0                 1.0  ...   
4    52.000000  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761  69.193991  25.778712       0.0          1.0                 1.0  ...   
762  54.000000  19.132653       0.0          1.0                 0.0  ...   
763  67.962041  25.439019       0.0          1.0                 1.0  ...   
764  68.499715  25.584960       0.0          1.0                 1.0  ...   
765  68.416958  25.564308       0.0          1.0                 1.0  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0    0.000000  0.000000               0.0                 0.236035   
1    0.000000  0.000000               0.0                 4.742909   
2    0.048191  0.087855               0.0                 5.852626   
3    0.000000  0.000000               0.0                 2.672416   
4    0.000000  0.000000               0.0                 1.000000   
..        ...       ...               ...                      ...   
761  0.074925 -0.003216               0.0                 4.272397   
762  0.000000  0.000000               0.0                 6.702640   
763  0.000000  0.000000               0.0                 5.093694   
764  0.065328  0.111087               0.0                 4.345900   
765  0.064508  0.119566               0.0                 4.373343   

     M6_Mental_health_support  m6_illness_events0_1_2  M6_drugs0_1  \
0                         0.0                     1.0          0.0   
1                         0.0                     2.0          0.0   
2                         0.0                     1.0          0.0   
3                         0.0                     1.0          0.0   
4                         0.0                     2.0          0.0   
..                        ...                     ...          ...   
761                       0.0                     1.0          0.0   
762                       0.0                     2.0          0.0   
763                       0.0                     2.0          0.0   
764                       0.0                     1.0          0.0   
765                       0.0                     1.0          0.0   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          1.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          1.0   
..                         ...                          ...   
761     

In [28]:
df['M0_employment_status'] = df['M0_employment_status'].abs()
df['M0_drinking'] = df['M0_drinking'].abs()
df['M0_diet'] = df['M0_diet'].abs()
df['M0_MedDiet'] = df['M0_MedDiet'].abs()
df['M0_SpecialDiet'] = df['M0_SpecialDiet'].abs()
df['M0_obese'] = df['M0_obese'].abs()
df['M0_exercise'] = df['M0_exercise'].abs()
df['M0_life_events_1_2'] = df['M0_life_events_1_2'].abs()
df['m0_menopausalstatuspre'] = df['m0_menopausalstatuspre'].abs()
df['M0_psychotropics'] = df['M0_psychotropics'].abs()
df['m0_Global_QLQ30groups65gl'] = df['m0_Global_QLQ30groups65gl'].abs()
df['m0_Phys_Fun_QLQ30groups80gl'] = df['m0_Phys_Fun_QLQ30groups80gl'].abs()
df['m3_life_events_1_2'] = df['m3_life_events_1_2'].abs()
df['m3_employer_support'] = df['m3_employer_support'].abs()
df['m3_domestic_help'] = df['m3_domestic_help'].abs()
df['M3_psychotropics'] = df['M3_psychotropics'].abs()
df['M3_Mental_health_support'] = df['M3_Mental_health_support'].abs()
df['m3_illness_events0_1_2'] = df['m3_illness_events0_1_2'].abs()
df['m3_Global_QLQ30groups65gl'] = df['m3_Global_QLQ30groups65gl'].abs()
df['m3_Phys_Fun_QLQ30groups80gl'] = df['m3_Phys_Fun_QLQ30groups80gl'].abs()
df['M6_employment_status'] = df['M6_employment_status'].abs()
df['M6_exercise'] = df['M6_exercise'].abs()
df['M6_life_events_1_2'] = df['M6_life_events_1_2'].abs()
df['M6_well_being_activities'] = df['M6_well_being_activities'].abs()
df['M6_domestic_help'] = df['M6_domestic_help'].abs()
df['M6_partner_reducework'] = df['M6_partner_reducework'].abs()
df['M6_psychotropics'] = df['M6_psychotropics'].abs()
df['M6_Mental_health_support'] = df['M6_Mental_health_support'].abs()
df['m6_illness_events0_1_2'] = df['m6_illness_events0_1_2'].abs()
df['M6_drugs0_1'] = df['M6_drugs0_1'].abs()
df['m6_Global_QLQ30groups65gl'] = df['m6_Global_QLQ30groups65gl'].abs()
df['m6_Phys_Fun_QLQ30groups80gl'] = df['m6_Phys_Fun_QLQ30groups80gl'].abs()
df['m18_Global_QLQ30groups65gl'] = df['m18_Global_QLQ30groups65gl'].abs()
df['m18_Global_QLQ30groups65gl'] = df['m18_Global_QLQ30groups65gl'].abs()

In [29]:
#create dataset based on the 78 most discriminating features (plus 'm6_Global_QLQ30groups65gl','m0_Global_QLQ30groups65gl','m18_Global_QLQ30groups65gl')
df = df[['M6_ptsd_PCL5','m3_sick_leave_days','M6_general_se_1_item','M6_sick_leave_days','M6_Pain_QLQ30',
         'M6_fear_of_recur_FCRI','M0_Diarrhoea_QLQ30','M0_Role_Fun_QLQ30','M0_resilience_CDRISC',
         'M6_Diarrhoea_QLQ30','M0_Financial_QLQ30','M3_Phys_Fun_QLQ30','M6_Upset_Hair_Image_br23','M3_Fatigue_QLQ30',
         'M6_distress_thermometer','M0_Breast_Symptoms_BR23','m6_performancestatus','M3_Constipation_QLQ30',
         'M6_mental_health_total_HADS','M3_Pain_QLQ30','m3_performancestatus',
         'M0_manageability_SOC','M6_positive_affect_PANAS','M6_Body_Image_br23','M6_employment_status',
         'm6_weight','M6_Phys_Fun_QLQ30','M6_partner_reducework',
         'M6_Arm_Symptoms_br23','M6_chemo','M0_mental_health_total_HADS','M6_Global_QLQ30','M0_Emot_Fun_QLQ30',
         'm3_domestic_help','M0_rumination_CERQ','M0_acceptance_CERQ',
         'M3_Breast_Symptoms_m3_br23','M6_Soc_Fun_QLQ30','M0_positive_affect_PANAS','M0_Upset_Hair_Image_BR23',
         'M3_Future_Persp_Image_m3_br23','M6_how_much_are_you_back_to_yourself','M6_Apetite_QLQ30','M3_Sex_Enjoy_m3_br23',
         'M3_Diarrhoea_QLQ30','M0_radio','M3_depression_HADS','M6_well_being_activities','M6_negative_affect_PANAS',
         'M3_surgery','M3_Dyspnoea_QLQ30','M6_anxiety_HADS',
         'M0_perspective_CERQ','m3_illness_events0_1_2','M0_Body_Image_BR23','M0_Forward_PACT','M6_Financial_QLQ30',
         'm0_menopausalstatuspre','M0_sick_leave_days','M6_Sex_Enjoy_br23',
         'M3_Side_Effects_m3_br23','m3_Global_QLQ30groups65gl','M0_openess_TIPI','m3_nccn_distress_thermometer','M6_Fatigue_QLQ30',
         'M3_Emot_Fun_QLQ30','M0_Arm_Symptoms_BR23','M3_Role_Fun_QLQ30','M0_conscientiousness_TIPI','M0_chemo',
         'M3_Apetite_QLQ30','M3_social_support_total',
         'M0_comprehensibility_SOC','M0_anxiety_HADS','M0_pos_refus_CERQ','M6_Breast_Symptoms_br23','M0_Sex_Funct_BR23',
         'm6_illness_events0_1_2','m6_Global_QLQ30groups65gl','m0_Global_QLQ30groups65gl','m18_Global_QLQ30groups65gl']]

In [30]:
df

M6_ptsd_PCL5  m3_sick_leave_days  M6_general_se_1_item  \
0       16.853781           37.903552             10.000000   
1       23.000000           46.773901              8.000000   
2       18.631052            0.000000              8.004340   
3        0.000000           20.972919             10.000000   
4        8.000000           45.226508              9.000000   
..            ...                 ...                   ...   
761     29.000000           29.204021              5.300000   
762     21.562097           70.000000              7.532526   
763     18.000000           90.000000              8.000000   
764     18.137143           36.425609              7.435147   
765     18.280142           36.215770              7.416486   

     M6_sick_leave_days  M6_Pain_QLQ30  M6_fear_of_recur_FCRI  \
0             22.170301      16.666667               0.000000   
1             50.705482      16.666667               0.625000   
2              5.539941      22.071444               1.283401   
3              3.872143       0.000000               0.777778   
4             41.062962      33.333333               1.777778   
..                  ...            ...                    ...   
761           26.767050      21.276956               1.851493   
762           80.000000      36.328864               1.222222   
763           55.291102       0.000000               2.333333   
764           29.939286      22.031478               1.719659   
765           29.726236      22.259999               1.731138   

     M0_Diarrhoea_QLQ30  M0_Role_Fun_QLQ30  M0_resilience_CDRISC  \
0              0.000000         100.000000              3.000000   
1              0.000000          83.333333              3.700000   
2              0.000000         100.000000              3.400000   
3              0.000000         100.000000              4.000000   
4              0.000000         100.000000              2.700000   
..                  ...                ...                   ...   
761            8.095402          80.523122              2.334360   
762            0.000000           0.000000              3.100000   
763            6.486714          82.726865              1.604264   
764            6.102001          79.498314              2.790543   
765            6.367531          79.441628              2.767817   

     M6_Diarrhoea_QLQ30  ...  M3_social_support_total  \
0              0.000000  ...                 4.375000   
1              0.000000  ...                 4.750000   
2              4.076183  ...                 4.061436   
3              0.000000  ...                 5.000000   
4              0.000000  ...                 4.000000   
..                  ...  ...                      ...   
761            8.199443  ...                 4.375000   
762            9.898142  ...                 4.061442   
763            0.000000  ...                 3.500000   
764            7.369380  ...                 4.061442   
765            7.466604  ...                 4.061447   

     M0_comprehensibility_SOC  M0_anxiety_HADS  M0_pos_refus_CERQ  \
0                   18.000000         0.285714           4.500000   
1                   23.000000         0.571429           2.500000   
2                   22.000000         0.142857           4.500000   
3                   25.000000         0.000000           5.000000   
4                   23.000000         0.285714           4.000000   
..                        ...              ...                ...   
761                 25.000000         0.941322           2.000000   
762                 27.000000         0.428571           2.500000   
763                 19.882465         1.571429           2.921420   
764                 20.700456         0.986603           3.019826   
765                 20.518714         0.989805           3.003943   

     M6_Breast_Symptoms_br23  M0_Sex_Funct_BR23  m6_illness_events0_1_2  \
0                  25.000000           0.000000              

------------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [31]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [32]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df.iloc[:,1:-1].values
Y=df.iloc[:,-1].values

In [33]:
classifier = SVC(kernel='rbf')

In [34]:
scores = []
cv = KFold(n_splits=10, random_state=42, shuffle=True)
for train_index, test_index in cv.split(X):
    X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]
    classifier.fit(X_train, Y_train)
    scores.append(classifier.score(X_test, Y_test))
ac_svm=np.mean(cross_val_score(classifier, X, Y, cv=10))
print(np.mean(cross_val_score(classifier, X, Y, cv=10)))

c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\sklearn\model_selection\_split.py:676: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=10.
  warnings.warn(
c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\sklearn\model_selection\_split.py:676: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=10.
  warnings.warn(


0.7572453861927546


<font color='red'>Accuracy: 75.72%</font>

In [35]:
y_predict = classifier.predict(X_test)

In [36]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0  1.0
row_0          
0.0     44    6
1.0     16   10

For 0: <font color='blue'>88%</font> (44/50) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>39%</font> (10/26) correctly classified -> <font color='red'>poor QoL</font> accuracy

------------

<u><font color='blue'>SMOTE</u></font>

In [37]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [38]:
from sklearn.datasets import make_classification
X, y = make_classification(n_samples=5000, n_features=2, n_informative=2,
                           n_redundant=0, n_repeated=0, n_classes=2,
                           n_clusters_per_class=1,
                           weights=[0.01, 0.05],
                           class_sep=0.8, random_state=0)
from imblearn.over_sampling import RandomOverSampler
ros = RandomOverSampler(random_state=0)
X_resampled, y_resampled = ros.fit_resample(X, y)
from collections import Counter
print(sorted(Counter(y_resampled).items()))

[(0, 2601), (1, 2601)]


In [40]:
X_train_ros, Y_train_ros = ros.fit_resample(X_train.astype('float'),Y_train)

In [41]:
#number of '0' and '1' after oversampling
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_ros))

Before SMOTE:  Counter({0.0: 451, 1.0: 238, 2.0: 1})
After SMOTE:  Counter({0.0: 451, 1.0: 451, 2.0: 451})


In [42]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_ros,Y_train_ros)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7105263157894737


<font color='red'>Accuracy: 71,05%</font>

In [43]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     38   12
1.0     10   16

For 0: <font color='blue'>76%</font> (38/50) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>62%</font> (16/26) correctly classified -> <font color='red'>poor QoL</font> accuracy

---------------

<font color='blue'><b><u>without</b></u></font> MICE Data Imoutation <br>
replacing remaining NaN values with the corresponding <u>mean value</u>

In [66]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   NaN          NaN      NaN         NaN             NaN   
762                   4.0          1.0      0.0         0.0             0.0   
763                   NaN          NaN      NaN         NaN             NaN   
764                   NaN          NaN      NaN         NaN             NaN   
765                   NaN          NaN      NaN         NaN             NaN   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0         77.0  28.628792       0.0          1.0                 1.0  ...   
1         52.0  22.506925       0.0          2.0                 1.0  ...   
2         68.0  27.942143       0.0          2.0                 1.0  ...   
3         53.0  24.196494       0.0          2.0                 1.0  ...   
4         52.0  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN       NaN          NaN                 1.0  ...   
762       54.0  19.132653       0.0          1.0                 0.0  ...   
763        NaN        NaN       NaN          NaN                 1.0  ...   
764        NaN        NaN       NaN          NaN                 NaN  ...   
765        NaN        NaN       NaN          NaN                 NaN  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0         0.0       0.0               0.0                      NaN   
1         0.0       0.0               0.0                      NaN   
2         NaN       NaN               0.0                      NaN   
3         0.0       0.0               0.0                      NaN   
4         0.0       0.0               0.0                      1.0   
..        ...       ...               ...                      ...   
761       NaN       NaN               0.0                      NaN   
762       0.0       0.0               0.0                      NaN   
763       0.0       0.0               0.0                      NaN   
764       NaN       NaN               NaN                      NaN   
765       NaN       NaN               NaN                      NaN   

     M6_Mental_health_support  m6_illness_events0_1_2  M6_drugs0_1  \
0                         0.0                     1.0          0.0   
1                         0.0                     2.0          0.0   
2                         NaN                     NaN          NaN   
3                         NaN                     1.0          0.0   
4                         0.0                     2.0          0.0   
..                        ...                     ...          ...   
761                       NaN                     NaN          0.0   
762                       0.0                     NaN          0.0   
763                       0.0                     NaN          0.0   
764                       NaN                     NaN          NaN   
765                       NaN                     NaN          NaN   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          1.0                          0.0   
2                          NaN                          NaN   
3                          0.0                          0.0   
4                          0.0                          1.0   
..                         ...                          ...   
761     

In [67]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 186 columns):
 #    Column                                                       Dtype  
---   ------                                                       -----  
 0    M0_employment_status                                         float64
 1    M0_drinking                                                  float64
 2    M0_diet                                                      float64
 3    M0_MedDiet                                                   float64
 4    M0_SpecialDiet                                               float64
 5    M0_weight                                                    float64
 6    M0_bmi                                                       float64
 7    M0_obese                                                     float64
 8    M0_exercise                                                  float64
 9    M0_life_events_1_2                                           fl

<b>- replace NaN values with mean of each column -</b>

In [68]:
df_mean = df.fillna(df.mean())

In [69]:
df_mean

M0_employment_status  M0_drinking   M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                 0.00000     1.000000  1.000000    1.000000        0.000000   
1                 0.00000     1.000000  1.000000    1.000000        0.000000   
2                 4.00000     1.000000  1.000000    1.000000        0.000000   
3                 2.00000     2.000000  1.000000    1.000000        0.000000   
4                 2.00000     1.000000  0.000000    0.000000        0.000000   
..                    ...          ...       ...         ...             ...   
761               1.32672     1.183024  0.781333    0.646667        0.137333   
762               4.00000     1.000000  0.000000    0.000000        0.000000   
763               1.32672     1.183024  0.781333    0.646667        0.137333   
764               1.32672     1.183024  0.781333    0.646667        0.137333   
765               1.32672     1.183024  0.781333    0.646667        0.137333   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792  0.000000     1.000000            1.000000  ...   
1    52.000000  22.506925  0.000000     2.000000            1.000000  ...   
2    68.000000  27.942143  0.000000     2.000000            1.000000  ...   
3    53.000000  24.196494  0.000000     2.000000            1.000000  ...   
4    52.000000  19.100092  0.000000     1.000000            1.000000  ...   
..         ...        ...       ...          ...                 ...  ...   
761  68.737012  25.580845  0.174625     1.034713            1.000000  ...   
762  54.000000  19.132653  0.000000     1.000000            0.000000  ...   
763  68.737012  25.580845  0.174625     1.034713            1.000000  ...   
764  68.737012  25.580845  0.174625     1.034713            0.632928  ...   
765  68.737012  25.580845  0.174625     1.034713            0.632928  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0     0.00000  0.000000          0.000000                  4.69246   
1     0.00000  0.000000          0.000000                  4.69246   
2     0.06483  0.100324          0.000000                  4.69246   
3     0.00000  0.000000          0.000000                  4.69246   
4     0.00000  0.000000          0.000000                  1.00000   
..        ...       ...               ...                      ...   
761   0.06483  0.100324          0.000000                  4.69246   
762   0.00000  0.000000          0.000000                  4.69246   
763   0.00000  0.000000          0.000000                  4.69246   
764   0.06483  0.100324          0.183432                  4.69246   
765   0.06483  0.100324          0.183432                  4.69246   

     M6_Mental_health_support  m6_illness_events0_1_2  M6_drugs0_1  \
0                    0.000000                 1.00000     0.000000   
1                    0.000000                 2.00000     0.000000   
2                    0.148265                 1.48731     0.048387   
3                    0.148265                 1.00000     0.000000   
4                    0.000000                 2.00000     0.000000   
..                        ...                     ...          ...   
761                  0.148265                 1.48731     0.000000   
762                  0.000000                 1.48731     0.000000   
763                  0.000000                 1.48731     0.000000   
764                  0.148265                 1.48731     0.048387   
765                  0.148265                 1.48731     0.048387   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                     0.000000                        1.000   
1                     1.000000                        0.000   
2                     0.222222                        0.415   
3                     0.000000                        0.000   
4                     0.000000                        1.000   
..                         ...                          ...

<b>- roundup values</b>  -

In [70]:
df_mean.M0_employment_status = df_mean.M0_employment_status.round()
df_mean.M0_drinking = df_mean.M0_drinking.round()
df_mean.M0_diet = df_mean.M0_diet.round()
df_mean.M0_MedDiet = df_mean.M0_MedDiet.round()
df_mean.M0_SpecialDiet = df_mean.M0_SpecialDiet.round()
df_mean.M0_obese = df_mean.M0_obese.round()
df_mean.M0_exercise = df_mean.M0_exercise.round()
df_mean.M0_life_events_1_2 = df_mean.M0_life_events_1_2.round()
df_mean.m0_menopausalstatuspre = df_mean.m0_menopausalstatuspre.round()
df_mean.M0_psychotropics = df_mean.M0_psychotropics.round()
df_mean.m0_Global_QLQ30groups65gl = df_mean.m0_Global_QLQ30groups65gl.round()
df_mean.m0_Phys_Fun_QLQ30groups80gl = df_mean.m0_Phys_Fun_QLQ30groups80gl.round()
df_mean.m3_life_events_1_2 = df_mean.m3_life_events_1_2.round()
df_mean.m3_employer_support = df_mean.m3_employer_support.round()
df_mean.m3_domestic_help = df_mean.m3_domestic_help.round()
df_mean.M3_psychotropics = df_mean.M3_psychotropics.round()
df_mean.M3_Mental_health_support = df_mean.M3_Mental_health_support.round()
df_mean.m3_illness_events0_1_2 = df_mean.m3_illness_events0_1_2.round()
df_mean.m3_Global_QLQ30groups65gl = df_mean.m3_Global_QLQ30groups65gl.round()
df_mean.m3_Phys_Fun_QLQ30groups80gl = df_mean.m3_Phys_Fun_QLQ30groups80gl.round()
df_mean.M6_employment_status = df_mean.M6_employment_status.round()
df_mean.M6_exercise = df_mean.M6_exercise.round()
df_mean.M6_life_events_1_2 = df_mean.M6_life_events_1_2.round()
df_mean.M6_well_being_activities = df_mean.M6_well_being_activities.round()
df_mean.M6_domestic_help = df_mean.M6_domestic_help.round()
df_mean.M6_partner_reducework = df_mean.M6_partner_reducework.round()
df_mean.M6_psychotropics = df_mean.M6_psychotropics.round()
df_mean.M6_Mental_health_support = df_mean.M6_Mental_health_support.round()
df_mean.m6_illness_events0_1_2 = df_mean.m6_illness_events0_1_2.round()
df_mean.M6_drugs0_1 = df_mean.M6_drugs0_1.round()
df_mean.m6_Global_QLQ30groups65gl = df_mean.m6_Global_QLQ30groups65gl.round()
df_mean.m6_Phys_Fun_QLQ30groups80gl = df_mean.m6_Phys_Fun_QLQ30groups80gl.round()
df_mean.m18_Global_QLQ30groups65gl = df_mean.m18_Global_QLQ30groups65gl.round()

In [71]:
df_mean

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   1.0          1.0      1.0         1.0             0.0   
762                   4.0          1.0      0.0         0.0             0.0   
763                   1.0          1.0      1.0         1.0             0.0   
764                   1.0          1.0      1.0         1.0             0.0   
765                   1.0          1.0      1.0         1.0             0.0   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792       0.0          1.0                 1.0  ...   
1    52.000000  22.506925       0.0          2.0                 1.0  ...   
2    68.000000  27.942143       0.0          2.0                 1.0  ...   
3    53.000000  24.196494       0.0          2.0                 1.0  ...   
4    52.000000  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761  68.737012  25.580845       0.0          1.0                 1.0  ...   
762  54.000000  19.132653       0.0          1.0                 0.0  ...   
763  68.737012  25.580845       0.0          1.0                 1.0  ...   
764  68.737012  25.580845       0.0          1.0                 1.0  ...   
765  68.737012  25.580845       0.0          1.0                 1.0  ...   

     M6_chemo  M6_radio  M6_psychotropics  M6_number_consultations  \
0     0.00000  0.000000               0.0                  4.69246   
1     0.00000  0.000000               0.0                  4.69246   
2     0.06483  0.100324               0.0                  4.69246   
3     0.00000  0.000000               0.0                  4.69246   
4     0.00000  0.000000               0.0                  1.00000   
..        ...       ...               ...                      ...   
761   0.06483  0.100324               0.0                  4.69246   
762   0.00000  0.000000               0.0                  4.69246   
763   0.00000  0.000000               0.0                  4.69246   
764   0.06483  0.100324               0.0                  4.69246   
765   0.06483  0.100324               0.0                  4.69246   

     M6_Mental_health_support  m6_illness_events0_1_2  M6_drugs0_1  \
0                         0.0                     1.0          0.0   
1                         0.0                     2.0          0.0   
2                         0.0                     1.0          0.0   
3                         0.0                     1.0          0.0   
4                         0.0                     2.0          0.0   
..                        ...                     ...          ...   
761                       0.0                     1.0          0.0   
762                       0.0                     1.0          0.0   
763                       0.0                     1.0          0.0   
764                       0.0                     1.0          0.0   
765                       0.0                     1.0          0.0   

     m6_Global_QLQ30groups65gl  m6_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          1.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          1.0   
..                         ...                          ...   
761     

In [72]:
#create dataset based on the 78 most discriminating features (plus 'm6_Global_QLQ30groups65gl','m0_Global_QLQ30groups65gl','m18_Global_QLQ30groups65gl')
df = df_mean[['M6_ptsd_PCL5','m3_sick_leave_days','M6_general_se_1_item','M6_sick_leave_days','M6_Pain_QLQ30',
         'M6_fear_of_recur_FCRI','M0_Diarrhoea_QLQ30','M0_Role_Fun_QLQ30','M0_resilience_CDRISC',
         'M6_Diarrhoea_QLQ30','M0_Financial_QLQ30','M3_Phys_Fun_QLQ30','M6_Upset_Hair_Image_br23','M3_Fatigue_QLQ30',
         'M6_distress_thermometer','M0_Breast_Symptoms_BR23','m6_performancestatus','M3_Constipation_QLQ30',
         'M6_mental_health_total_HADS','M3_Pain_QLQ30','m3_performancestatus',
         'M0_manageability_SOC','M6_positive_affect_PANAS','M6_Body_Image_br23','M6_employment_status',
         'm6_weight','M6_Phys_Fun_QLQ30','M6_partner_reducework',
         'M6_Arm_Symptoms_br23','M6_chemo','M0_mental_health_total_HADS','M6_Global_QLQ30','M0_Emot_Fun_QLQ30',
         'm3_domestic_help','M0_rumination_CERQ','M0_acceptance_CERQ',
         'M3_Breast_Symptoms_m3_br23','M6_Soc_Fun_QLQ30','M0_positive_affect_PANAS','M0_Upset_Hair_Image_BR23',
         'M3_Future_Persp_Image_m3_br23','M6_how_much_are_you_back_to_yourself','M6_Apetite_QLQ30','M3_Sex_Enjoy_m3_br23',
         'M3_Diarrhoea_QLQ30','M0_radio','M3_depression_HADS','M6_well_being_activities','M6_negative_affect_PANAS',
         'M3_surgery','M3_Dyspnoea_QLQ30','M6_anxiety_HADS',
         'M0_perspective_CERQ','m3_illness_events0_1_2','M0_Body_Image_BR23','M0_Forward_PACT','M6_Financial_QLQ30',
         'm0_menopausalstatuspre','M0_sick_leave_days','M6_Sex_Enjoy_br23',
         'M3_Side_Effects_m3_br23','m3_Global_QLQ30groups65gl','M0_openess_TIPI','m3_nccn_distress_thermometer','M6_Fatigue_QLQ30',
         'M3_Emot_Fun_QLQ30','M0_Arm_Symptoms_BR23','M3_Role_Fun_QLQ30','M0_conscientiousness_TIPI','M0_chemo',
         'M3_Apetite_QLQ30','M3_social_support_total',
         'M0_comprehensibility_SOC','M0_anxiety_HADS','M0_pos_refus_CERQ','M6_Breast_Symptoms_br23','M0_Sex_Funct_BR23',
         'm6_illness_events0_1_2','m6_Global_QLQ30groups65gl','m0_Global_QLQ30groups65gl','m18_Global_QLQ30groups65gl']]

In [73]:
df

M6_ptsd_PCL5  m3_sick_leave_days  M6_general_se_1_item  \
0       17.923599           37.903297             10.000000   
1       23.000000           37.903297              8.000000   
2       17.923599            0.000000              7.470353   
3        0.000000           37.903297             10.000000   
4        8.000000           37.903297              9.000000   
..            ...                 ...                   ...   
761     29.000000           37.903297              5.300000   
762     17.923599           70.000000              7.470353   
763     18.000000           90.000000              8.000000   
764     17.923599           37.903297              7.470353   
765     17.923599           37.903297              7.470353   

     M6_sick_leave_days  M6_Pain_QLQ30  M6_fear_of_recur_FCRI  \
0             32.032927      16.666667               0.000000   
1             32.032927      16.666667               0.625000   
2             32.032927      22.131148               1.699881   
3             32.032927       0.000000               0.777778   
4             32.032927      33.333333               1.777778   
..                  ...            ...                    ...   
761           32.032927      22.131148               1.699881   
762           80.000000      22.131148               1.222222   
763           32.032927       0.000000               2.333333   
764           32.032927      22.131148               1.699881   
765           32.032927      22.131148               1.699881   

     M0_Diarrhoea_QLQ30  M0_Role_Fun_QLQ30  M0_resilience_CDRISC  \
0              0.000000         100.000000              3.000000   
1              0.000000          83.333333              3.700000   
2              0.000000         100.000000              3.400000   
3              0.000000         100.000000              4.000000   
4              0.000000         100.000000              2.700000   
..                  ...                ...                   ...   
761            6.023551          79.601990              2.783082   
762            0.000000           0.000000              3.100000   
763            6.023551          79.601990              2.783082   
764            6.023551          79.601990              2.783082   
765            6.023551          79.601990              2.783082   

     M6_Diarrhoea_QLQ30  ...  M3_social_support_total  \
0              0.000000  ...                 4.375000   
1              0.000000  ...                 4.750000   
2              7.255865  ...                 4.061444   
3              0.000000  ...                 5.000000   
4              0.000000  ...                 4.000000   
..                  ...  ...                      ...   
761            7.255865  ...                 4.375000   
762            7.255865  ...                 4.061444   
763            0.000000  ...                 3.500000   
764            7.255865  ...                 4.061444   
765            7.255865  ...                 4.061444   

     M0_comprehensibility_SOC  M0_anxiety_HADS  M0_pos_refus_CERQ  \
0                   18.000000         0.285714           4.500000   
1                   23.000000         0.571429           2.500000   
2                   22.000000         0.142857           4.500000   
3                   25.000000         0.000000           5.000000   
4                   23.000000         0.285714           4.000000   
..                        ...              ...                ...   
761                 25.000000         0.984750           2.000000   
762                 27.000000         0.428571           2.500000   
763                 20.538807         1.571429           3.010246   
764                 20.538807         0.984750           3.010246   
765                 20.538807         0.984750           3.010246   

     M6_Breast_Symptoms_br23  M0_Sex_Funct_BR23  m6_illness_events0_1_2  \
0                  25.000000           0.000000              

------------------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [74]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [75]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df.iloc[:,1:-1].values
Y=df.iloc[:,-1].values

In [76]:
classifier = SVC(kernel='rbf')

In [77]:
scores = []
cv = KFold(n_splits=10, random_state=42, shuffle=True)
for train_index, test_index in cv.split(X):
    X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]
    classifier.fit(X_train, Y_train)
    scores.append(classifier.score(X_test, Y_test))
ac_svm=np.mean(cross_val_score(classifier, X, Y, cv=10))
print(np.mean(cross_val_score(classifier, X, Y, cv=10)))

0.7571599453178401


<font color='red'>Accuracy: 75.72%</font>

In [79]:
y_predict = classifier.predict(X_test)

In [80]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0  1.0
row_0          
0.0     54    1
1.0     17    4

For 0: <font color='blue'>98%</font> (54/55) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>19%</font> (4/21) correctly classified -> <font color='red'>poor QoL</font> accuracy

-------------

<u><font color='blue'>SMOTE</u></font>

In [81]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [82]:
from sklearn.datasets import make_classification
X, y = make_classification(n_samples=5000, n_features=2, n_informative=2,
                           n_redundant=0, n_repeated=0, n_classes=2,
                           n_clusters_per_class=1,
                           weights=[0.01, 0.05],
                           class_sep=0.8, random_state=0)
from imblearn.over_sampling import RandomOverSampler
ros = RandomOverSampler(random_state=0)
X_resampled, y_resampled = ros.fit_resample(X, y)
from collections import Counter
print(sorted(Counter(y_resampled).items()))

[(0, 2601), (1, 2601)]


In [83]:
X_train_ros, Y_train_ros = ros.fit_resample(X_train.astype('float'),Y_train)

In [84]:
#number of '0' and '1' after oversampling
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_ros))

Before SMOTE:  Counter({0.0: 517, 1.0: 173})
After SMOTE:  Counter({0.0: 517, 1.0: 517})


In [85]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_ros,Y_train_ros)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.6447368421052632


<font color='red'>Accuracy: 64,47%</font>

In [90]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     36   19
1.0     10   11

For 0: <font color='blue'>66%</font> (36/55) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>52%</font> (11/21) correctly classified -> <font color='red'>poor QoL</font> accuracy

------------------